### DISCUSSION TITLE
Explore alternatives to Laravel

### DISCUSSION THREAD

**Discussion by chuckadams** [aspirepress/AspireCloud#67] [Ideas]
The only thing that made Laravel's abuse of magic methods tolerable was `barryvdh/laravel-ide-helper` and it doesn't work for Laravel 11.5 and higher, as it gets confused by the generics and generates nonexistent types like `\Illuminate\Database\Eloquent\TValue`.  barryvdh/laravel-ide-helper#1572 is closed, but the same issue remains even when using the dev-master version of the helper.  

We need to consider the pros and cons of our framework options.  I'm listing some below, but please respond with your own suggestions, whether it's a new item in the list or a new bullet point for an existing item.

### Laravel

* **Arguments in favor**
  * Large community
  * Large ecosystem

* **Arguments against**
  * Abuse of magic methods makes most static analysis impossible
  * The many pitfalls of ActiveRecord, and in particular Eloquent's implementation of it
  * laravel-ide-helper has regressed to the point of being no longer usable

### Symfony

* **Arguments in favor**
  * Highly mature ecosystem, used as the base of other frameworks, including Laravel
  * Statically checked configuration: a typo in a config key is an error, not ignored
  * Powerful DI container allowing for autoconfiguration, decorators, and tagged services
  * Doctrine ORM keeps magic confined to autogenerated proxies
  * Attributes can express lots of logic declaratively, from routing to validation to serialization 

* **Arguments against**
  * Relatively small community, unless you're French 🇫🇷 
  * yaml config gets hairy (yaml is not required, but it's pretty standard)
  * Data mapper ORM (Doctrine) has less syntax sugar, demands using repositories
  * ArrayCollection is highly anemic compared Laravel collections (illuminate/collections does work but Doctrine doesn't use it)

### Mezzio

* **Arguments in favor**
  * Clean design built on PSR-7 middleware encourages non-mutating patterns
  * Large menu of options to mix and match DI containers, routing, and view layers.

* **Arguments against**
  * Documentation is extremely scant.
  * Default configuration requires wiring up factories and services by hand.
  * DIY approach limits interoperability of components to impractically narrow interfaces like PSR-11

**Comment by sarah-savage** [comment]
I understand taht developer experience is extremely important when working with the code. Given the size of the project and the scope of the effort, I imagine that my day-to-day will be more focused on managing the effort than getting my hands in the code. So I am sensitive to and cognizant of any effort to improve the developer expereince.

Where I become concerned is when we have shifted framework already in the last week from Mezzio to something new (Laravel), which was the consensus of people writing code on the project. We're at high risk here of pursuing perfection in the form of the tool, rather than pursuing function in the form of the output.

I'd like to see if we can get feedback from someone like @taylorotwell who I will email about how to set up the IDE ideally for a Laravel project. I'll report back. In the meantime, feel free to catalog the ideals that you're looking for, and we'll compare notes on Friday.

**Comment by Tmeister** [comment]
Hi all,

As I said on Slack, having this conversation now is better than jumping around and doing the same work three times.

Here is my list based on my experience. **On a personal note**, I have never worked with Mezzio and barely used Symfony.

## Laravel

### Pros
* Extensive ecosystem with well-maintained packages
* Outstanding documentation and large community support
* Rich built-in features (auth, queue, jobs, broadcasting)
* Eloquent ORM with intuitive syntax
* Artisan CLI for scaffolding and management
* Great for rapid application development
* Easy learning curve for beginners
* Regular release schedule
* Strong community packages
* Built-in testing tools

### Cons
* Heavy reliance on magic methods (Agree)
* IDE support issues (especially with newer versions)
* Higher memory footprint
* Dependency on specific Laravel ways of doing things (Agree)

## Symfony

### Pros
* Enterprise-grade stability and security
* Excellent dependency injection
* Strong typing and IDE support
* High performance
* Modular component system
* Mature ecosystem
* Highly customizable

### Cons
* Steeper learning curve
* More verbose configuration
* Slower development for simple apps
* Smaller community than Laravel
* More boilerplate code required
* Complex configuration system
* Can be overwhelming for beginners

## Mezzio
*(From what I read, TBH, I didn't know this framework):*

### Pros
* Extremely lightweight
* PSR-7 and PSR-15 compliant
* High performance
* Minimal overhead
* Great for APIs

### Cons
* Smaller community
* Mainly unknown?
* Limited out-of-box functionality
* More initial setup required
* Fewer ready-made solutions
* Limited documentation
* More architectural decisions needed
* Steeper learning curve for beginners
* Less third-party packages
* More time needed for common features

Now, talking about the specific issue you are seeing, I think this is fixed: https://github.com/barryvdh/laravel-ide-helper/pull/1591

My vote is for Laravel due to the ecosystem around it, the easy learning curve, and the fact that we want (I think) more people to be able to jump in. Symfony and Mezzio put the bar too high, IMO.

Rust and Go was just a joke.

**Comment by chuckadams** [comment]
I assure you that the specific issue with laravel-ide-helper I referred to is not fixed, not in the 3.2 release or dev-master.  Check out the `feat/ac-64/new-tables` branch and run `make helpers`, see for yourself.  It generates gems like this in _ide_helper.php:

```
             * @template TValue
             * @param (\Closure(): TValue)|\Illuminate\Database\Eloquent\list<string> $columns
             * @param (\Closure(): TValue)|null $callback
             * @return \Illuminate\Database\Eloquent\TModel|\Illuminate\Database\Eloquent\TValue 
             * @static 
             */            public static function firstOr($columns = [], $callback = null)
            {
                                /** @var \Illuminate\Database\Eloquent\Builder $instance */
                                return $instance->firstOr($columns, $callback);
            }
```

Seeing the problem with the `@return` clause?  It can't really even be fixed with sed, the TModel parameter isn't even declared.  But my bigger point is that a mature framework should not have to require such a script just to be able to autocomplete and avoid "method not found" errors for its own core APIs.

**Comment by asirota** [comment]
What about this as a pro/con to each : experience of the current AspireCloud volunteers with each framework?

I mean experience counts here and the framework with the deepest level of experience would be the one to go with. 

Sounds like laravel is that system, yeah?

Would it not be true that we can influence the laravel team to try to fix the problems that we find? Esp if we get enough momentum.

**Comment by taylorotwell** [comment]
Hey there! If there is something we can do in Laravel itself to address the IDE Helper issues let me know or send a PR anytime! 😊

Are you all mainly using PHP Storm? Are you using their "Laravel Idea" plug-in?

You all may also be interested to know we are releasing our own free first-party VS Code extension in November. Demoed here: https://www.youtube.com/watch?v=AwWepVU5uWM&t=485s

**Reply by sarah-savage** [reply]
Hi Taylor. Thanks for your input. 

I am using PHPStorm and I have the Idea plugin. I’ll let @chuckadams speak to his specific issues but my goal is for us to learn the ways developers use Laravel to be productive. We need to ship, and fast. 

I appreciate your willingness to comment here and be involved. Once @chuckadams has a moment I’m sure he’ll let you know what the relevant issues are (and what IDE he’s using). 

Thanks again!

**Reply by chuckadams** [reply]
@taylorotwell I'd love to get to the bottom of the apparent regression with barryvdh/laravel-ide-helper, but I also know it's neither your circus nor monkeys, so I'll just have to maintain phpdoc by hand for the moment.  I'm normally loath to use paid plugins, but now I can see the Laravel IDEA plugin is worth more than every penny of its low price _(don't tell them, they'll jack up the price!)_ 😉 

I could bend your ear about the philosophical objections I have to ActiveRecord and Eloquent's implementation of it in particular.  I used to hack on Class::DBI 15 years back or so, so I could share some stories on ORMs.  But in the end, while I'm still firmly in the Symfony "camp" as preferences go, I've used Laravel since the 5.8 era and found it a workhorse, even if a somewhat cantankerous one.

**Reply by sarah-savage** [reply]
Thanks @chuck. And thank you @taylorotwell for your willingness to support us and spend your time with us. 

I’m glad we found a resolution and can move forward.

**Comment by chuckadams** [comment]
So conclusion to this 24-hour tizzy of mine: I still maintain that without proper IDE tooling, Laravel is difficult to work with safely, to the point of being a complete deal-breaker.  The completely free tool (barryvdh/laravel-ide-helper) still has issues with generics in Laravel 11.5+, but there is now readily available IDE tooling that's either free or cheap as chips.  

So, while I still prefer Symfony's overall declarative approach, I withdraw the motion to change frameworks.  Laravel it stays.

**Reply by sarah-savage** [reply]
Your motion to withdraw is noted, and hearing no objections granted.

**Comment by pleveris** [comment]
Hello,

As this is a first comment hof mine on This project here, it goes with a short intro...
The idea looks very interesting! I'm not a developer of WordPress (?anymore? :D), but the whole idea seems a challenging and inspiring!

Refering to the discussion object, As a Back-End PHP developer who has experience in both Laravel and Symfony, I can just confirm that Laravel is way easier to learn for beginners, while a good point for Symfony is that they do use strict coding standards and coding convensions :)
In case of the IDE, I personally am using VSCode; so as Taylor mentioned in a comment, am so excited to try their new extension once it comes out! :)
As this project already got my attention, will be following it from now on; Hope to become a contributor (one or other way) in a near future!

**Reply by chuckadams** [reply]
Glad to see new faces around here, [come on over to our Slack](https://aspirepress.org/slack/) and hang out if you like!
